libs/traits/src/lib.rs

```diff
@@ -774,40 +544,61 @@ pub trait TokenSwaps<Account> {
 	///     buy_amount: 15 * FOREIGN_ASSET_0_DECIMALS,
 	///     initial_buy_amount: 100 * FOREIGN_ASSET_0_DECIMALS,
 	///     sell_rate_limit: Quantity::checked_from_integer(2u32).unwrap(),
-	///     min_fullfillment_amount: 6 * FOREIGN_ASSET_0_DECIMALS,
+	///     min_fulfillment_amount: 6 * FOREIGN_ASSET_0_DECIMALS,
 	///     max_sell_amount: 30 * FOREIGN_ASSET_1_DECIMALS
 	/// }
 	fn update_order(
 		account: Account,
 		order_id: Self::OrderId,
 		buy_amount: Self::Balance,
 		sell_rate_limit: Self::SellRatio,
-		min_fullfillment_amount: Self::Balance,
+		min_fulfillment_amount: Self::Balance,
 	) -> DispatchResult;
 
 	/// Cancel an already active order.
 	fn cancel_order(order: Self::OrderId) -> DispatchResult;
 
 	/// Check if the order is still active.
 	fn is_active(order: Self::OrderId) -> bool;
+
+	/// Check whether there already exist orders for the given trading pair.
+	fn order_pair_exists(currency_in: Self::CurrencyId, currency_out: Self::CurrencyId) -> bool;
+
+	/// Check whether there already exist counter orders for the given trading
+	/// pair against which could be traded.
+	fn counter_order_pair_exists(
+		currency_in: Self::CurrencyId,
+		currency_out: Self::CurrencyId,
+	) -> bool;
 }
 
-/// Trait to handle Investment Portfolios for accounts
-pub trait InvestmentsPortfolio<Account> {
-	type InvestmentId;
-	type CurrencyId;
+/// Trait to transmit a change of status for anything uniquely identifiable.
+///
+/// NOTE: The main use case to handle asynchronous operations.
+pub trait StatusNotificationHook {
+	/// The identifying type
+	type Id;
+	/// The type for possible states
+	type Status;
+	/// The error type
+	type Error: Debug;
+
+	/// Notify that the status has changed for the given id
+	fn notify_status_change(id: Self::Id, status: Self::Status) -> Result<(), Self::Error>;
+}
+
+/// Trait to synchronously provide a currency conversion estimation for foreign
+/// currencies into/from pool currencies.
+pub trait SimpleCurrencyConversion {
+	type Currency;
 	type Balance;
 	type Error;
-	type AccountInvestmentPortfolio;
-
-	/// Get the payment currency for an investment.
-	fn get_investment_currency_id(
-		investment_id: Self::InvestmentId,
-	) -> Result<Self::CurrencyId, Self::Error>;
-
-	/// Get the investments and associated payment currencies and balances for
-	/// an account.
-	fn get_account_investments_currency(
-		who: &Account,
-	) -> Result<Self::AccountInvestmentPortfolio, Self::Error>;
+
+	/// Estimate the worth of an outgoing currency amount in the incoming
+	/// currency.
+	fn stable_to_stable(
+		currency_out: Self::Currency,
+		amount_out: Self::Balance,
+		currency_in: Self::Currency,
+	) -> Result<Self::Balance, Self::Error>;
```

Comment:
I do not want to add noise with this, but maybe it's interesting in order to prove things work well at compile time that `amount_out` and the retuning `Balance` be different types. Something like: `BalanceOut` and `BalanceIn`. These types will be propagated to `foreign-investment`.  The only way from `foreign-investment` perspective to mix these values is to call `stable_to_stable` before making the conversion and then ensuring at compile time that two balances with different representation are not computed together.

Reply:
I would put this on the _subsequent PR_ list as in practice both `BalanceIn` and `BalanceOut` would just be `Balance`.

Reply:
Sure! 👍🏻